### PR TITLE
uefi-capsule: Add support for CapsuleOnDisk

### PR DESCRIPTION
**Pull request by hughsie**
Based on a patch by Ilias Apalodimas <ilias.apalodimas@linaro.org>, many thanks.

Fixes https://github.com/fwupd/fwupd/issues/2900

Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [X] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/uefi-capsule/fu-plugin-uefi-capsule.c
Comment (hughsie) [inline plugins/uefi-capsule/fu-plugin-uefi-capsule.c:18]: @superm1 maybe fu-uefi-disk-device.h might be clearer? Or more unclear perhaps?

File: plugins/uefi-capsule/fu-uefi-cod-device.c
Comment (apalos) [inline plugins/uefi-capsule/fu-uefi-cod-device.c:103]: you need to read OsIndications before seeting it and OR it

File: plugins/uefi-capsule/fu-uefi-cod-device.c
Comment (apalos) [inline plugins/uefi-capsule/fu-uefi-cod-device.c:130]: if the efivarfs is mounted as r/o we need to ignore the setvariable call and just copy the capsule.  
U-Boot has a config option that ignores OsIndications before applying the update.
On the next reboot the Capsule0000 efi variable will be missing, so that's our hint the update didn't run at all.

File: plugins/uefi-capsule/fu-uefi-backend-linux.c
Comment (apalos) [inline plugins/uefi-capsule/fu-uefi-backend-linux.c:100]: Why do we need to copy the value here? Can't we just use buf?

File: plugins/uefi-capsule/fu-plugin-uefi-capsule.c
Comment (superm1) [reply plugins/uefi-capsule/fu-plugin-uefi-capsule.c:18]: cod is fine

File: plugins/uefi-capsule/fu-uefi-backend-linux.c
Comment (superm1) [inline plugins/uefi-capsule/fu-uefi-backend-linux.c:83]: inverting a negative we get from the config file makes this confusing, can we just set the disable instead?

File: plugins/uefi-capsule/fu-uefi-cod-device.c
Comment (superm1) [inline plugins/uefi-capsule/fu-uefi-cod-device.c:36]: seems like a broad assumption, would it make more sense to store the variable name that was used into the history database or something?

File: plugins/uefi-capsule/fu-uefi-cod-device.c
Comment (superm1) [inline plugins/uefi-capsule/fu-uefi-cod-device.c:58]: this seems to be asking for a struct definition of what this header format is supposed to be looking like.  then you can do the check with `sizeof(CapsuleResultVariableHeader)` and access from `header->guid` and such.

File: plugins/uefi-capsule/fu-uefi-cod-device.c
Comment (apalos) [reply plugins/uefi-capsule/fu-uefi-cod-device.c:36]: You can do easier than that. 
Start a loop and append the 'i' to Capsule.  So you'll read
Capsule0001
Capsule0002 etc, 
Once you fail to read a variable, you know your parsing needs to stop

File: plugins/uefi-capsule/fu-uefi-cod-device.c
Comment (superm1) [inline plugins/uefi-capsule/fu-uefi-cod-device.c:54]: if you already verified the size is big enough to contain the GUID, do you really need the same memory copy function?

File: plugins/uefi-capsule/fu-uefi-cod-device.c
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-cod-device.c:103]: Fixed, thanks.

File: plugins/uefi-capsule/fu-uefi-backend-linux.c
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-backend-linux.c:100]: I was wrong, memcpy isn't endian-clean. Fixed.

File: plugins/uefi-capsule/fu-uefi-backend-linux.c
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-backend-linux.c:83]: whoops, that inversion was a typo.

File: plugins/uefi-capsule/fu-uefi-cod-device.c
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-cod-device.c:58]: That's not endian clean tho; although I admit in this context LE is probably the only thing we care about.

File: plugins/uefi-capsule/fu-uefi-cod-device.c
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-cod-device.c:54]: No, but is saves the security team worrying that it's a random memcpy. i.e. for ~3 integer comparisons we don't even need to think about "is this safe" even if we remove the total length ready in the future.

File: plugins/uefi-capsule/fu-uefi-cod-device.c
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-cod-device.c:130]: I've added this check.

File: plugins/uefi-capsule/fu-uefi-cod-device.c
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-cod-device.c:36]: I've added something like this; can you verify it does the right thing? We might need to add another error in the `g_error_matches()` block.

Comment (apalos) [reply plugins/uefi-capsule/fu-uefi-cod-device.c]: @hughsie, sure I'll try to run the whole thing tomorrow, but it seems to be doing the right thing

Comment (apalos) [reply plugins/uefi-capsule/fu-uefi-cod-device.c]: You can also read the CapsuleLast variable (which will contain the name of the last applied capsule e.g Capsule0001) and figure out the size of 'idx'. The later is more 'efish' but both will work.

I did manage to run the whole thing and the ro/rw scanning of efivarfs is working correctly. The capsules are installed and executed and Capsule0000 is creted.  Here's the get-history output 

`Loading…                 [-                                      ]07:01:34:0315 FuEngine             failed to add device /sys/devices/platform/emmc2bus/fe340000.emmc2/mmc_host/mmc0/mmc0:0007/block/mmcb )
Loading…                 [***************************************]
WARNING: This package has not been validated, it may not work properly.
Unknown Product
│
└─Unknown Firmware:
  │   Device ID:          a45df35ac0e948ee180fe216a5f703f32dda163f
  │   Previous version:   0
  │   Update State:       Success
  │   Last modified:      2021-07-23 06:41
  │   GUID:               e2bb9c06-70e9-4b14-97a3-5a7913176e3f
  │   Device Flags:       • Internal device
  │                       • Updatable
  │                       • System requires external power source
  │                       • Needs a reboot after installation
  │                       • Device is usable for the duration of the update
  │ 
  └─  New version:      1.0.0
        License:          Unknown
`

**Comment by apalos** [comment]
@hughsie one of the things that I missed on my initial patchset due to my setup is the BootNext potential usage.  The EFI spec says 
`The directory \EFI\UpdateCapsule is checked for capsules only within the EFI system partition on the device specified in the active boot option determine by reference to BootNext variable or BootOrdervariable processing. The active Boot Variable is the variable with highest priority BootNext or within BootOrder that refers to a device found to be present. Boot variables in BootOrder but referring to devices not present are ignored when determining active boot variable. `

So what this means is if you Boot####(0000, 0001 etc) variables point to the ESP partition that contains the 'EFI/Updatecapsule' directory, then your capsule update on-disk will run on next reboot and we don't have to do anything.  However if the Boot#### variables point outside that directory, you need to create a Boot#### variable that points to it (even if if doesn't contain any executable) and set BootNext to that boot variable.

Given the SetVariable limitations in U-Boot, I think we need to document that.  It's up to you if you want to add the extra functionality, but we could always do it in a follow up patch?

File: plugins/uefi-capsule/fu-uefi-cod-device.c
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-cod-device.c]: CapsuleLast contains the string Capsule0001? Could you please upload a tarball with the exact Capsule0001 and CapsuleLast contents and I'll add them to git and add some tests.

File: plugins/uefi-capsule/fu-uefi-cod-device.c
Comment (apalos) [inline plugins/uefi-capsule/fu-uefi-cod-device.c:39]: FU_EFIVAR_GUID_EFI_GLOBAL is not valid for Capsule#### we'll need EFI_CAPSULE_REPORT_GUID (from the EFI spec)

File: plugins/uefi-capsule/fu-uefi-cod-device.c
Comment (apalos) [inline plugins/uefi-capsule/fu-uefi-cod-device.c:103]: The CapsuleLast will have the last capsule we updated, so instead of running the loop from 0 here, we must read CapsuleLast (which will have somelike like Capsule0002, and start from 2

File: plugins/uefi-capsule/fu-uefi-cod-device.c
Comment (apalos) [inline plugins/uefi-capsule/fu-uefi-cod-device.c:102]: There's a r/o CapsuleMax variable we ca read

File: plugins/uefi-capsule/fu-uefi-cod-device.c
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-cod-device.c:38]: Fixed, thanks.

File: plugins/uefi-capsule/fu-uefi-cod-device.c
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-cod-device.c]: I generated some for the tests.

**Comment by apalos** [comment]
> #define STRLEN(s) (sizeof(s)/sizeof(s[0]) -1) would work?

But if your concern here is code generation, just use sizeof() instead of strlen() and it will be computed on compile time

**Comment by hughsie** [comment]
@apalos I think this addresses most of your concerns, right?

**Comment by hughsie** [comment]
@superm1 this also has the advantage that you can test deploying a capsule using either CoD, grub or shim using fwupdtool now.

**Comment by apalos** [comment]
> @apalos I think this addresses most of your concerns, right?

Yea, looks good on a first look.  I'll try testing it within the week and let you know

File: plugins/uefi-capsule/fu-plugin-uefi-capsule.c
Comment (superm1) [inline plugins/uefi-capsule/fu-plugin-uefi-capsule.c:728]: nit: In the scenario it's not supported, this message will emit `failed to check CapsuleOnDisk support: Capsule-on-Disk is not supported` which seems wrong.
It should be instead `not using CapsuleOnDisk support: %s`

File: plugins/uefi-capsule/fu-uefi-grub-device.c
Comment (superm1) [inline plugins/uefi-capsule/fu-uefi-grub-device.c:2]: 2021, and I think I was part of this one too

File: plugins/uefi-capsule/fu-uefi-grub-device.c
Comment (superm1) [inline plugins/uefi-capsule/fu-uefi-grub-device.c:91]: do you think we need this debugging still?  I was pretty useful early on, but I think we only really care about logging for failures.

File: plugins/uefi-capsule/fu-uefi-grub-device.c
Comment (superm1) [inline plugins/uefi-capsule/fu-uefi-grub-device.c:95]: maybe just return `return g_spawn_sync`?

File: plugins/dell/fu-self-test.c
Comment (superm1) [inline plugins/dell/fu-self-test.c:556]: why a new variable?  can't we just flag the existing `FWUPD_UEFI_TEST`?

File: plugins/uefi-capsule/fu-uefi-backend.c
Comment (superm1) [inline plugins/uefi-capsule/fu-uefi-backend.c:77]: I think this should be `FU_TYPE_UEFI_NVRAM_DEVICE`.  If secure boot is turned off it will not be using shim.  Shim is only used for SB.

File: plugins/uefi-capsule/fu-uefi-cod-device.c
Comment (superm1) [inline plugins/uefi-capsule/fu-uefi-cod-device.c:34]: comment seems wrong?

File: plugins/uefi-capsule/fu-uefi-cod-device.c
Comment (superm1) [inline plugins/uefi-capsule/fu-uefi-cod-device.c:36]: rather than attempting to read, how about just prefix the error of `failed to read %s`

File: plugins/uefi-capsule/fu-uefi-tool.c
Comment (superm1) [inline plugins/uefi-capsule/fu-uefi-tool.c:173]: update to `nvram` instead of `shim` if you make the other changes

File: plugins/uefi-capsule/fu-uefi-tool.c
Comment (superm1) [inline plugins/uefi-capsule/fu-uefi-tool.c:172]: Maybe when presenting this it should be "device update method"

File: plugins/uefi-capsule/fu-plugin-uefi-capsule.c
Comment (hughsie) [reply plugins/uefi-capsule/fu-plugin-uefi-capsule.c:728]: Fixed, ta.

File: plugins/uefi-capsule/fu-uefi-grub-device.c
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-grub-device.c:2]: I'm really sorry, I've added you first in that file.

File: plugins/uefi-capsule/fu-uefi-grub-device.c
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-grub-device.c:91]: I've hidden it with a FWUPD_UPDATE_CAPSULE_VERBOSE

File: plugins/uefi-capsule/fu-uefi-grub-device.c
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-grub-device.c:95]: Fixed.

File: plugins/uefi-capsule/fu-uefi-cod-device.c
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-cod-device.c:34]: Fixed

File: plugins/uefi-capsule/fu-uefi-cod-device.c
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-cod-device.c:36]: Fixed.

File: plugins/uefi-capsule/fu-uefi-tool.c
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-tool.c:172]: Fixed, thanks.

File: plugins/uefi-capsule/fu-uefi-backend.c
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-backend.c:77]: Agreed, changed.

File: plugins/uefi-capsule/fu-uefi-tool.c
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-tool.c:173]: Fixed.

File: plugins/dell/fu-self-test.c
Comment (hughsie) [reply plugins/dell/fu-self-test.c]: Agreed, fixed.

File: plugins/uefi-capsule/fu-uefi-grub-device.c
Comment (superm1) [inline plugins/uefi-capsule/fu-uefi-grub-device.c:161]: I think you want these to only run in non self tests (IE move them later in the function).

**Review by superm1 — DISMISSED** [review]
Just one problem, and then LGTM.  It would be good to make a call out to the OEMs to test master in the various scenarios once this is merged.

File: plugins/uefi-capsule/fu-uefi-grub-device.c
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-grub-device.c:161]: Or move the tests for FWUPD_UEFI_TEST up? Do you think we want to run fu_uefi_device_write_update_info() and fu_uefi_device_fixup_firmware() for example?

**Comment by hughsie** [comment]
I/we/you actually need to test nvram, grub and @apalos needs to test CoD on this branch; but I'm away from home until Thurs and don't want a broken laptop just yet :)

**Comment by superm1** [comment]
> I/we/you actually need to test nvram, grub and @apalos needs to test CoD on this branch; but I'm away from home until Thurs and don't want a broken laptop just yet :)

With the change I pushed (and various other changes I've opened PR's for with unrelated issues) it works for GRUB on a system for me.

**Review by superm1 — DISMISSED** [review]
LGTM and GRUB mode tests well for me now.

File: plugins/uefi-capsule/fu-uefi-grub-device.c
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-grub-device.c:165]: Fixed, thanks.

**Comment by apalos** [comment]
@hughsie on a quick pass this seems fine. I'll test it on monday and give you an update!

**Pull request merged**
